### DISCUSSION TITLE
[Question] Since main.go is generated, where should initialization logic be performed? i.e. setting up a database

### DISCUSSION THREAD

**Discussion by donovanrost** [livebud/bud#146] [Q&A] [answered]

**Comment by matthewmueller** [comment, marked as answer]
Hey @donovanrost! The automatic dependency injection should help you with initializing your dependencies (like a database) on boot. The nice thing about this is that you will _(soon)_ be able to do this with other entrypoints like tests so there's much less initialization boilerplate.

I wrote an example in [this comment](https://github.com/livebud/bud/discussions/72#discussioncomment-2788198) that I think is relevant here. 

Create an `env/env.go` file:

```go
package env

import "github.com/caarlos0/env"

// Load the environment
func Load() (*Env, error) {
  var e Env
  if err := env.Parse(&env); err != nil {
    return err
  }
  return &e, nil
}


type Env struct {
  DatabaseURL string `env:DATABASE_URL`
}
```

Then in a `db/db.go` file:

```go
package db

import "app.com/env"
import "github.com/jackc/pgx"

func Dial(env *env.Env) (*Client, error) {
  return pgx.Dial(env.DatabaseURL)
}

type Client = pgx.Client
```

Finally in your `controller/controller.go`:

```go
package controller

import "app.com/db"

type Controller struct {
  DB *db.Client
}

type User struct {
  Name string
}

func (c *Controller) Index() ([]*User, error) {
  rows, err := c.DB.Query("select * from users")
  if err != nil {
    return nil, err
  }
  // TODO: turn rows into users
  return users, nil
}
```

With that you can now switch out your development database with your production database using:

```sh
DATABASE_URL="..." bud run
```

Let me know how that works for you! We still may need to be able to hook into `main.go`, but I think this use case is covered.

**Reply by matthewmueller** [reply]
Marking as the answer, but let me know if you have any more questions about this @donovanrost!
